File: source/python/deque_pointers.py
# Double-ended queue using a doubly-linked list

class Node:
    def __init__(self, val):
        self.val = val
        self.prev = None
        self.next = None

dq_head = None
dq_tail = None
dq_size = 0

def push_front(val):
    global dq_head, dq_tail, dq_size
    node = Node(val)
    if dq_head is None:
        dq_head = node
        dq_tail = node
    else:
        node.next = dq_head
        dq_head.prev = node
        dq_head = node
    dq_size += 1
    print(f"push_front: {val}")

def push_back(val):
    global dq_head, dq_tail, dq_size
    node = Node(val)
    if dq_tail is None:
        dq_head = node
        dq_tail = node
    else:
        node.prev = dq_tail
        dq_tail.next = node
        dq_tail = node
    dq_size += 1
    print(f"push_back: {val}")

def pop_front():
    global dq_head, dq_tail, dq_size
    if dq_head is None:
        print("deque empty")
        return None
    val = dq_head.val
    nxt = dq_head.next
    if nxt is not None:
        nxt.prev = None
    else:
        dq_tail = None
    dq_head = nxt
    dq_size -= 1
    print(f"pop_front: {val}")
    return val

def pop_back():
    global dq_head, dq_tail, dq_size
    if dq_tail is None:
        print("deque empty")
        return None
    val = dq_tail.val
    prv = dq_tail.prev
    if prv is not None:
        prv.next = None
    else:
        dq_head = None
    dq_tail = prv
    dq_size -= 1
    print(f"pop_back: {val}")
    return val

def print_deque():
    parts = []
    cur = dq_head
    while cur is not None:
        parts.append(str(cur.val))
        cur = cur.next
    print("deque: " + " ".join(parts))

push_front(1)
push_front(2)
push_back(3)
push_back(4)
print_deque()
print(f"size: {dq_size}")

pop_front()
pop_back()
print_deque()
print(f"size: {dq_size}")

while dq_head is not None:
    pop_front()
print("deque cleared")

File: source/go/deque_pointers.go
package main

import "fmt"

type DequeNode struct {
	val        int
	prev, next *DequeNode
}

var dqHead *DequeNode
var dqTail *DequeNode
var dqSize = 0

func pushFront(val int) {
	node := &DequeNode{val: val}
	if dqHead == nil {
		dqHead = node
		dqTail = node
	} else {
		node.next = dqHead
		dqHead.prev = node
		dqHead = node
	}
	dqSize++
	fmt.Printf("push_front: %d\n", val)
}

func pushBack(val int) {
	node := &DequeNode{val: val}
	if dqTail == nil {
		dqHead = node
		dqTail = node
	} else {
		node.prev = dqTail
		dqTail.next = node
		dqTail = node
	}
	dqSize++
	fmt.Printf("push_back: %d\n", val)
}

func popFront() (int, bool) {
	if dqHead == nil {
		fmt.Println("deque empty")
		return 0, false
	}
	node := dqHead
	val := node.val
	nxt := node.next
	if nxt != nil {
		nxt.prev = nil
	} else {
		dqTail = nil
	}
	dqHead = nxt
	dqSize--
	fmt.Printf("pop_front: %d\n", val)
	return val, true
}

func popBack() (int, bool) {
	if dqTail == nil {
		fmt.Println("deque empty")
		return 0, false
	}
	node := dqTail
	val := node.val
	prv := node.prev
	if prv != nil {
		prv.next = nil
	} else {
		dqHead = nil
	}
	dqTail = prv
	dqSize--
	fmt.Printf("pop_back: %d\n", val)
	return val, true
}

func printDeque() {
	fmt.Print("deque:")
	cur := dqHead
	for cur != nil {
		fmt.Printf(" %d", cur.val)
		cur = cur.next
	}
	fmt.Println()
}

func main() {
	pushFront(1)
	pushFront(2)
	pushBack(3)
	pushBack(4)
	printDeque()
	fmt.Printf("size: %d\n", dqSize)

	popFront()
	popBack()
	printDeque()
	fmt.Printf("size: %d\n", dqSize)

	for dqHead != nil {
		popFront()
	}
	fmt.Println("deque cleared")
}
